### DISCUSSION TITLE
Difference between queryCache.clear() vs queryCache.refetchQueries()

### DISCUSSION THREAD

**Discussion by transmissions11** [TanStack/query#1231] [Q&A] [answered]
When should I be using `queryCache.clear()` vs `queryCache.refetchQueries()`?

Does `queryCache.clear()` just clear the cache or refetch as well?

If I have a user switch accounts, should I run both `queryCache.refetchQueries()` and `queryCache.clear()` or just one or the other?

**Comment by TkDodo** [comment, marked as answer]
Hm, I haven't used either. I mostly do `invalidateQueries`.

Tanner answered the question regarding the difference between `refetchQueries` and `invalidateQueries` like this:

![Screenshot 2020-10-30 at 20 28 38](https://user-images.githubusercontent.com/1021430/97748985-9c321a80-1aee-11eb-94e2-6bea2edb6615.png)

> If I have a user switch accounts, should I run both queryCache.refetchQueries() and queryCache.clear() or just one or the other?

I'm pretty sure that `queryCache.clear()` just removes the value from the cache, which I think is what you want when a user switches accounts. Once the user visits a page again where the query is observed, it will fetch again because it has no data. If you are certain that the data will be needed immediately, you could `refetch` it right away.

Alternatively, you could also put the `accountId` _into_ the query key. That way, you would automatically re-fetch if the account changes, and the unused queries would be garbage collected after `cacheTime` has passed, which defaults to 5 minutes.

**Reply by transmissions11** [reply]
Thank you for your thorough answer! 

I actually lied a tiny bit (for simplicity) when I said I was wondering for when a user logs out, the use case I'm actually thinking about right now is when a user deposits and then their account balance increases, I didn't know what method to use. I think I'll go with `invalidateQueries` or `refetchQueries`. I might end up going with `refetchQueries` because the account balance data is needed throughout almost the entire app and effects multiple queries so I don't want any stale data showing up to the user or queries refetching as they open a modal or something and they see $0 for a second.

Thank you again!

**Reply by weyert** [reply]
The image says:
> Tanner Linsley: invalidateQueries is a bit smarter (with some different defaults) to avoid overmatching e.g. not refetching inactive queries and only marking them as stale
> Tanner Linsley: refetchQueries has none of that, so it will for sure refetch the query if it matches, even if it's inactive
